lapce-ui/src/editor.rs

```diff
@@ -1782,6 +1785,124 @@ impl LapceEditor {
         }
     }
 
+    /// Checks if the cursor is on a bracket and highlights the matching bracket if there is one.
+    /// If the cursor is between brackets it highlights the enclosing brackets.
+    fn paint_bracket_highlight(
+        ctx: &mut PaintCtx,
+        data: &LapceEditorBufferData,
+        screen_lines: &ScreenLines,
+    ) {
+        if screen_lines.lines.is_empty() {
+            return;
+        }
+
+        let cursor_offset = data.editor.cursor.offset();
+
+        let char_at_cursor = match data.doc.buffer().char_at_offset(cursor_offset) {
+            Some(c) => c,
+            None => return,
+        };
+
+        let syntax = match data.doc.syntax() {
+            Some(syntax) => syntax,
+            None => return,
+        };
+
+        if is_bracket(char_at_cursor) {
+            if let Some(new_offset) = syntax.find_matching_pair(cursor_offset) {
+                Self::highlight_char(ctx, data, screen_lines, cursor_offset);
+                Self::highlight_char(ctx, data, screen_lines, new_offset);
+            }
+        } else {
+            let end_line = *screen_lines.lines.last().unwrap();
```

Comment:
There's existing code for this logic as well.

See https://github.com/lapce/lapce/blob/master/lapce-data/src/document.rs#L2510

Essentially it tries to use tree-sitter to find a bracket first. They only accept a single bracket though, so you might want to extend those functions to work with an array of different brackets, and with some offset boundary as an option so that you only scan the things on your screen.